sqlobject/src/test/java/org/jdbi/v3/sqlobject/TestVariousOddities.java

```diff
@@ -151,4 +154,49 @@ public void testAbiguityResolved() {
         @Override
         String value();
     }
+
+    public interface OnDemandOddities extends SqlObject, Transactional<OnDemandOddities> {
+        @CreateSqlObject
+        VersionA versionA();
+    }
+
+    public class DecoratedOnDemandOddities implements OnDemandOddities {
+        private OnDemandOddities onDemand = dbRule.getJdbi().onDemand(OnDemandOddities.class);
+
+        @Override
+        public Handle getHandle() {
+            return onDemand.getHandle();
+        }
+
+        @Override
+        public <R, X extends Exception> R withHandle(HandleCallback<R, X> callback) throws X {
+            return onDemand.withHandle(callback);
+        }
+
+        @Override
+        public VersionA versionA() {
+            return onDemand.versionA();
+        }
+    }
+
+    @Test
+    public void onDemandCreateSqlObject() throws Exception {
+        assertThat(dbRule.getJdbi().onDemand(OnDemandOddities.class).versionA().value())
```

Comment:
Why should this work? You create an onDemand object and call a query method outside its demand scope with no outer scope to piggyback on.

Reply:
The short answer is, "because users expect it to".  `@CreateSqlObject` is a bit of a wart on the API.  Imagine that you have two types of SqlObject, "Type A" and "Type B".  (ondemand vs attached).  It's surprising to have the created sub-sql-object not be of the same type as the parent, and plenty of people have gotten tripped up.  Since there's a clearer semantic that is also actually more useful (a closed SqlObject is useless) we can remove the surprise.

Reply:
You want CreateSqlObject instances to inherit the parent's onDemand aspect?

Reply:
Yep.  See #1011 for background info.

Reply:
I find a fix for this a bit of anti-feature tbh. That CreateSqlObject subobjects are visible to callers is a wart indeed. Users should ignore them because they're _obviously_ not intended to be used independently of the parent onDemand object, they're essentially private state. If you want to use one of the CreateSqlObjected classes, you should create your own instance properly instead of piggybacking on "accidentally" exposed factory methods.

Making this work is an API-subverting convenience that sends a signal to users that violating the internal kitchen boundary of a class is an acceptable design to deliberately go with, rather than an unfortunate wart (interface components forcibly being public) that should be ignored.

Reply:
I don't disagree that `@CreateSqlObject` is a bit of a wart -- but there's no fewer than three different users that have been bit by the confusing and broken behavior (personally, we had a production outage due to the onDemand scope not working as expected) so I don't think "not fixing it" is really an option.  Perhaps we drop `@CreateSqlObject` entirely in v4 but as long as it lives in our tree we should try to make it behave as nicely as we can.  And the current behavior does nothing useful, and throws a confusing and difficult to diagnose error message.  This PR doesn't have to be the fix, but we must ship *some* fix for this issue.

Reply:
It's really only as confusing as hibernate entities being detached outside transactional methods... I find CreateSqlObject as it works now ok, you just have to refrain from thinking you can pull a subcomponent out of the owning class' scope and use it. Be it 4 or 40 users making that mistake, it's just not what the feature is for and there's no functional need to use it that way.

I won't turn this change down, just know I find it an anti-feature that encourages violating the internal kitchen boundary for a minor convenience, which is bad in its own right.

Reply:
I agree with your sentiment, but let's take this conversation to a larger "What should CreateSqlObject be (or not be) in 4.0" at a later time.  For now let's make the existing feature as least-warty as it can be.

Reply:
Well yeah, but that's just it... I find this change even more warty because it isn't even consistent with itself anymore.